manticore/core/cpu/disasm.py

```diff
@@ -0,0 +1,162 @@
+        pass
+
+    @abstractproperty
+    def size(self):
+        pass
+
+    @abstractproperty
+    def operands(self):
+        pass
+
+    # FIXME (theo) eliminate one of the two of insn_name, name
+    @abstractproperty
+    def insn_name(self):
+        pass
+
+    @abstractproperty
+    def name(self):
+        pass
+
+class Disasm(object):
+    """Abstact class for different disassembler interfaces"""
+
+    def __init__(self, disasm):
+        self.disasm = disasm
+
+    @abstractmethod
+    def disassemble_instruction(self, code, pc):
+        """Get next instruction based on the disassembler in use
+
+        :param code: disassembled code
+        :param pc: program counter
+        """
+        pass
+
+class CapstoneDisasm(Disasm):
+    def __init__(self, arch, mode):
+        cap = cs.Cs(arch, mode)
+        cap.detail = True
+        cap.syntax = 0
+        super(CapstoneDisasm, self).__init__(cap)
+
+    def disassemble_instruction(self, code, pc):
+        """Get next instruction based on Capstone disassembler
+
+        :param code: disassembled code
+        :param pc: program counter
+        """
+        return next(self.disasm.disasm(code, pc))
+
+class BinjaILDisasm(Disasm):
+
+    def __init__(self, view):
+        self.bv = view
+        # dictionary with llil for each function. This will be consumed
+        # using an iterator, so that we don't repeat ourselves whenever
+        # we ask for the next IL
+        self.func_llil = {}
+        super(BinjaILDisasm, self).__init__(view)
+
+    def disassemble_instruction(self, _, pc):
+        """Get next instruction based on Capstone disassembler
+
+        :param code: disassembled code
+        :param pc: program counter
+        """
+        #  print(self.bv.get_disassembly(pc))
+        blocks = self.bv.get_basic_blocks_at(pc)
+        func = blocks[0].function
+        fllil = self.func_llil.get(func,
+                                   iter([il for block in func.low_level_il
+                                         for il in block]))
+
+        il = next(fllil)
+        print(il)
+        print ("%s %x %x\n") % (il.operation.name, il.instr_index, il.address)
+        self.func_llil[func] = fllil
+        return self.BinjaILInstruction(il)
+
+
+    class BinjaILInstruction(Instruction):
+        def __init__(self, llil):
+            self.llil = llil
+            super(BinjaILDisasm.BinjaILInstruction, self).__init__()
+
+        @property
+        def size(self):
+            return 1
+
+        @property
+        def operands(self):
+            return self.llil.operands
+
+        @operands.setter
+        def operands(self, value):
+            self._operands = value
+
+        @property
+        def insn_name(self):
+            return self.llil.operation.name
+
+        @property
+        def name(self):
+            return self.llil.operation.name[len("LLIL_"):]
+
+class BinjaDisasm(Disasm):
+
+    def __init__(self, view):
+        self.bv = view
```

Comment:
I'm not sure that does what you think it does. That method is for appending LLIL to an existing function; it doesn't return an arbitrary LLIL instruction. In particular, it generates Lifted IL, not Low Level IL, and it appends it to the end of the `LowLevelILFunction` object passed to it, with no context. Here's an example, using that on an LLIL instruction `if (al != 0) then 12 @ 0x14000210 else 18 @ 0x140002105`:
```>>> a = LowLevelILFunction(Architecture['x86_64'])
>>> Architecture['x86_64'].get_instruction_low_level_il(bv.read(0x140002103, 8), 0x140002103, a)
2L
>>> a[0]
<il: if (!=) then 1 else 2>
>>> a[1]
<il: jump(0x140002110)>
>>> a[2]
<il: jump(0x140002105)>
```
Notice those are `LLIL_JUMP` instructions, rather than `LLIL_GOTO` instructions; the former are used when an instruction isn't in the current `LowLevelILFunction`, whereas the former is used when the target is in the current `LowLevelILFunction`. I'm not sure what the performance effects are on doing something like this, but it would probably be a little confusing, and you're also going to then have to track flags explicitly, since Lifted IL hasn't abstracted the flags away or folded conditions like LLIL does